### DISCUSSION TITLE
API Changes

### DISCUSSION THREAD

**Discussion by TokisanGames** [TokisanGames/Terrain3D#21] [Architecture]
This discussion can house threads on changes to the API, especially renaming things.

**Comment by TokisanGames** [comment]
**This Repository: GDExtensionTerrain**

* Rather than GDExtensionTerrain, I think Terrain3D is a better name. Especially since the existing Terrain3D presumably is going to be archived. 

* `addons/terrain` should then be renamed to `addons/terrain_3d`

What do you think?

**Comment by TokisanGames** [comment]
**TerrainStorage::maps**

* Maps should only refer to the height map and control map(s)
* `add/remove/has_map` should be called map_segment, section, or **region** so they aren't confused w/ full sized maps. I'd also just drop map and call it `add_region`

* `add_map(vector3)` should be `add_map_at(vector3)`. I use this conversion elsewhere and plan to put in things like `get_height_at(Vector3)`. So with both this and above, I'd name these three `add_region_at(Vector3)`
* `map_offsets` can be called `region_offsets`

These are my thoughts, what are yours?

If agreed, I can implement these and the other discussions below.

**Reply by outobugi** [reply]
I think 'add_region' looks prettier than 'add_region_at'. I know, stupid argument but I have an obsession to make function names look nice. 😆 
Otherwise I agree and I already started working on the API changes before you made this discussion.

**Reply by TokisanGames** [reply]
`add_region` suggests `add_region(Map/Image)` because we're adding a thing passed from the coder. Or `add_region()` if nothing is passed).

`add_region(Vector3)` suggests `add_region_at` because the function definition suggests the coder is not providing the region to add, but adding it at a location. And there might be an `add_region(map)` and `add_region_at(vec3)`.

I agree function names should look nice. But more importantly, they should be clear as to what they do. That is why I made this and all suggestions.

It is a minor point. If this hasn't convinced you, we can let it go for now.

**Reply by outobugi** [reply]
> If this hasn't convinced you, we can let it go for now.

:)

Like you said it's a minor point and I'd like to mention, to enforce my opinion, that e.g AABB class has 'expand' method which takes "to point".

**Comment by TokisanGames** [comment]
**Materials**

* TerrainLayerMaterial aren't layers really. That name is confusing. They are just materials. How to differentiate it from the single master terrain material? I would rename the master to that or global. People won't be interfacing with the master. They will be interfacing with the layers, so it should be clearer. TerrainMaterial -> `TerrainMasterMaterial` or `TerrainGlobalMaterial`, TerrainLayerMaterial -> `TerrainMaterial`. 

* `set_material(0, mat)`, `get_material(0)`, `get_materials`, `get_material_count` all make more sense to me than layers.

* `void set_layer(const Ref<TerrainLayerMaterial3D> &p_material, int p_index);` I think the index should come first, as is done in `set_surface_material(0, mat)` and elsewhere.

* Master `set_material` -> `set_master_material` or `set_global_material` or `set_terrain_material`
* material var changes to match above

* TerrainLayerMaterial::`albedo` and `set_albdeo`. Maybe these are better named `albedo_tint`. Also do we need this? IMO, we should drop this and include a color map so the user can paint various colors. We don't need a per texture color tint. If they want that, they should change the texture in photoshop.

**Comment by TokisanGames** [comment]
**TerrainStorage funcs & vars**

* `_update_arrays()` - will probably go away as UVs can be paintable and colors turn into a global map
*  `_update_textures()` better named `_update_terrain_textures()` or `_update_terrain_texture_arrays()`.
* `_update_layers()` -> `_update_materials()`
* `_update_material()` better as `_update_maps()` or `_update_terrain_maps()`

* `albedo_textures` -> `albedo_texture_array`, and normal
* In _update_textures, swap `albedo_textures` and `albedo_texture_array`. Each is more accurate of a name for the opposite. 
* `_convert_array()` -> `_create_texture_array`
